### PR TITLE
Reverse swipe-responsibility (opt-out -> opt-in) (DEV)

### PR DESCRIPTION
**Pull request by d4rken**
Follow up to #3209.
In that PR we wanted to make one item swipable, but enabling our previous swipe functionality, requires to explicitly disable swiping for any items that shouldn't be swiped, which was unexpected.

This PR reverses it. Swiping is "opt-in", not "opt-out".

Call `RecyclerView.setupSwipe()` on a RecyclerView, then any `ViewHolder` that implements `Swipeable` can be swiped, all others can NOT be swiped.

I've also refactored the responsibility for this to be more modular:
*The fragment does not have to be aware of any details of the swipe functionality.
 `RecyclerView.setupSwipe()` has a default callback implementation that checks if the touched `ViewHolder` implements `Swipeable`, and then calls the `Swipeable.onSwiped`.
* The `ViewHolder` can also override the default `val Swipeable.movementFlags: Int? get() = null` if it wants to decide to not move for some reason.
* The implementing `Swipeable` is similar to an `OnClickListener` and the responsibility of the UI (ViewHolder in this case), not the data item.

### Testing
* Swipe Checkins
* Swipe vaccination certs
* Swipe camera permission view

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/list/SwipeExtension.kt [hunk 1]
```diff
@@ -18,20 +18,15 @@ import kotlin.math.min
...
- * ```
+ * After calling this on a recyclerview, every ViewHolder that implements [Swipeable] is swipeable.
  */
-fun RecyclerView.onSwipeItem(
+fun RecyclerView.setupSwipe(
     context: Context,
-    onSwipe: (position: Int, direction: Int) -> Unit
+    onSwipe: (holder: RecyclerView.ViewHolder, direction: Int) -> Unit = { holder, direction ->
+        (holder as? Swipeable)?.let { holder.onSwipe(holder, direction) }
+    }
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/list/SwipeExtension.kt:29]: Important bit 2, how swipe functionality is enabled for a list.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/presencetracing/attendee/checkins/items/ActiveCheckInVH.kt [hunk 2]
```diff
@@ -20,7 +21,14 @@ class ActiveCheckInVH(parent: ViewGroup) :
...
         parent = parent
-    ) {
+    ),
+    Swipeable {
+
+    private var latestItem: Item? = null
+
+    override fun onSwipe(holder: RecyclerView.ViewHolder, direction: Int) {
+        latestItem?.let { it.onSwipeItem(it.checkin, holder.adapterPosition) }
+    }
```
[the remark this answers is not in the document]
Comment (mtwalli) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/presencetracing/attendee/checkins/items/ActiveCheckInVH.kt:31]: Two points against this specific part:
1- IDE is showing is not used , can someone delete it 
2- Holding the item in the Viewholder , View can be used in recycler view this means that many items can use the same view, right?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/list/Swipeable.kt [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    fun onSwipe(holder: RecyclerView.ViewHolder, direction: Int)
+}
+
+/**
+ * returns true if the item is as [Swipeable] , false otherwise
+ * and helps with smart cast
+ */
+
+@OptIn(ExperimentalContracts::class)
+fun HasStableId?.isSwipeable(): Boolean {
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/list/Swipeable.kt:34]: this is not used anywhere now, can be deleted

**Review by mtwalli — COMMENTED** [review]
I tested it, it works as expected, some comments need to be clarified

**Review by chris-cwa — DISMISSED** [review]
works on pixel 3 and 4

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/presencetracing/attendee/checkins/items/ActiveCheckInVH.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/presencetracing/attendee/checkins/items/ActiveCheckInVH.kt:31]: > 1- IDE is showing is not used , can someone delete it

Done

> 2- Holding the item in the Viewholder , View can be used in recycler view this means that many items can use the same view, right?

The ViewHolder is re-used, but not for multiple items at the time. When the list is scrolled and no longer visible viewholders are recycled, `onBind` will be called with the new item, before the viewholder is displayed again.

[the remark this answers is not in the document]
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/presencetracing/attendee/checkins/items/ActiveCheckInVH.kt:31]: Just to note here, there is an edge case where a no longer displayed viewholder is not yet recycled, could still keep a reference to `latestItem` even if the ViewModel or its original data source is already working with new data.

The impact is negligible though as the reference would only be kept until the viewholder is destroyed (which happens sometime after it is no longer visible and not being recycled) or the screen is changed. We could `null` the reference when the viewholder is no longer used, but I don't think it's worth it here, the data classes are small.

[the remark this answers is not in the document]
Comment (mtwalli) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/presencetracing/attendee/checkins/items/ActiveCheckInVH.kt:31]: Let us keep it simple, if it works. I'm OK with that.

**Review by mtwalli — APPROVED** [review]

**Pull request merged**